lib/spack/spack/relocate.py

```diff
@@ -467,39 +480,66 @@ def _replace_prefix_bin(filename, byte_prefixes):
     """Replace all the occurrences of the old install prefix with a
     new install prefix in binary files.
 
-    The new install prefix is prefixed with ``b'/'`` until the
-    lengths of the prefixes are the same.
+    The new install prefix is required to be at either:
+
+    1. At least 16 characters shorter than the original, or
+    2. Share the same 16-character suffix with the original
+
+    This is to avoid, with high probability, replacing a path suffix that's
+    getting re-used as the storage for another string constant.  For more
+    detail, see:
+    - https://github.com/spack/spack/pull/31739
+    - https://github.com/spack/spack/pull/32253
+    - TODO: add pr with this fix here
+
+    No longer prefixed with with ``os.sep`` until the lengths of the prefixes
+    are the same, instead the new prefix is terminated at its original length to
+    leave a suffix unmodified as much as possible.
 
     Args:
         filename (str): target binary file
         byte_prefixes (OrderedDict): OrderedDictionary where the keys are
-        binary strings of the old prefixes and the values are the new
-        binary prefixes
+        bytes representing the old prefixes and the values are the new
+        prefixes (all bytes utf-8 encoded)
     """
-    all_prefixes = re.compile(b"|".join(re.escape(prefix) for prefix in byte_prefixes.keys()))
-
-    def padded_replacement(old):
-        new = byte_prefixes[old]
-        pad = len(old) - len(new)
-        if pad < 0:
-            raise BinaryTextReplaceError(old, new)
-        return new + b"/" * pad
+    all_prefixes = re.compile(
+        b"(" + b"|".join(re.escape(prefix) for prefix in byte_prefixes.keys()) + b")\0?"
+    )
+    print(all_prefixes)
 
+    escaped_null = re.compile(re.escape(b"\0") + b"$")
     with open(filename, "rb+") as f:
-        # Register what replacement string to put on what offsets in the file.
-        replacements_at_offset = [
-            (padded_replacement(m.group(0)), m.start())
-            for m in re.finditer(all_prefixes, f.read())
-        ]
+        data = f.read()
+        f.seek(0)
+        for match in all_prefixes.finditer(data):
+            found_full = match.group(0)
+            old = escaped_null.sub(b"", found_full)
+            terminated = old != found_full
+
+            new = byte_prefixes[old]
 
-        # Apply the replacements
-        for replacement, offset in replacements_at_offset:
-            f.seek(offset)
-            f.write(replacement)
+            if len(new) > len(old) or (terminated and new[-7:] != old[-7:]):
```

Comment:
I'm confused, why require a common suffix? You get `len(old) - len(new) - 1` unused bytes after replacement + new null byte, so a new suffix is possible when that's >= 7. And if you require a common suffix, you can just as well pad with leading `/`, like in the non-null terminated case.

Reply:
Darn it, you're right, the check isn't quite right anymore.

The suffix case is mainly to deal with same-length or shorter than safe width difference prefixes when terminated.  We can't do anything if the suffixes aren't the same.  It also impacts padding in similar cases. Probably it should be:
* if len(new) - len(old)  > 7 and terminated: ok, shorten
* if len(new) - len(old)  > 7 and not terminated: ok, pad
* if len(new) - len(old)  < 7 and terminated and matching: ok, pad
* if len(new) - len(old)  < 7 and terminated and not matching: 💥 
* if len(new) - len(old)  < 7 and not terminated and matching: ok, pad
* if len(new) - len(old)  < 7 and not terminated and not matching: may be ok to pad, depends on length of suffix

Need to add tests so we cover all of these.  Honestly kinda want to fuzz the 💩 out of this too.

Reply:
I'm happy to push a few commits when you're on holidays ;p

Reply:
How about

```python
if len(new) > len(old):
  raise NoInplaceUpdatePossible()
elif old[-N:] == new[-N:]:  # common suffix, whether trailing null or not
  replacement = b"/" * (len(old) - len(new)) + new
elif len(new) + N < old:
  replacement = new + b"\0"
else:
  raise AliasingRiskIssues()
```

Reply:
Awesome! I'm floored @haampie, this looks fabulous.  I threw in one suggestion to fix an indentation bug but pending a defensive recheck on gitlab I think this is GTG.